### PR TITLE
Add youtube hls streaming wip

### PR DESCRIPTION
**Pull request by mv2731**
This pull request includes the changes made to support HLS ingestion streaming to YouTube. I added an obs_ouptut_info struct and registered it in order to make this possible and then wrote the corresponding methods for the struct. Additionally, I added parameters so that start and init functions could recognize the output format as HLS.

### CODE REVIEW

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 1]
```diff
@@ -354,6 +363,51 @@ static bool ffmpeg_mux_start(void *data)
...
+		return false;
+	}
+
+	/* write headers and start capture */
+	os_atomic_set_bool(&stream->active, true);
+	os_atomic_set_bool(&stream->capturing, true);
+	stream->total_bytes = 0;
+	obs_output_begin_data_capture(stream->output, 0);
+
+	info("Writing file '%s'...", stream->path.array);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Writing file -> Writing to path

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 2]
```diff
@@ -525,14 +527,19 @@ static int ffmpeg_mux_init_context(struct ffmpeg_mux *ffm)
 	AVOutputFormat *output_format;
 	int ret;
 	bool isNetwork = false;
+	bool isHLS = false;
 	if (strncmp(ffm->params.file, SRT_PROTO, sizeof(SRT_PROTO) - 1) == 0 ||
 	    strncmp(ffm->params.file, UDP_PROTO, sizeof(UDP_PROTO) - 1) == 0 ||
 	    strncmp(ffm->params.file, TCP_PROTO, sizeof(TCP_PROTO) - 1) == 0)
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Add a check for HTTP protocol here as well.

Comment (ushadow) [reply plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Then if isNetwork is true, call avformat_network_init().

If isNetwork is true and format is not set, guess format to be "mpegts", else if format is set, guess format to be format.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 3]
```diff
@@ -229,6 +236,8 @@ static void add_muxer_params(struct dstr *cmd, struct ffmpeg_muxer *stream)
 static void build_command_line(struct ffmpeg_muxer *stream, struct dstr *cmd,
 			       const char *path)
 {
+	if ((stream->format == NULL) || (*stream->format == '\0'))
+		stream->format = "";
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: I don't see where stream->format is set in the command line.

**Review by ushadow — CHANGES_REQUESTED** [review]
Please see the comments.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 4]
```diff
@@ -256,9 +256,7 @@ static bool init_params(int *argc, char ***argv, struct main_params *params,
 	}
 
 	*p_audio = audio;
-
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Don't remove the empty line here and the one below.

Comment (mv2731) [reply plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Done.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 5]
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
...
 #define UDP_PROTO "udp"
 #define TCP_PROTO "tcp"
+#define HTTP_PROTO "http"
 
 static int ffmpeg_mux_init_context(struct ffmpeg_mux *ffm)
 {
 	AVOutputFormat *output_format;
 	int ret;
 	bool isNetwork = false;
+	bool isHTTP = false;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: bool is_http = (strncmp(ffm->params.file, HTTP_PROTO,
			  sizeof(HTTP_PROTO) - 1) == 0);

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 6]
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 #define SRT_PROTO "srt"
 #define UDP_PROTO "udp"
 #define TCP_PROTO "tcp"
+#define HTTP_PROTO "http"
 
 static int ffmpeg_mux_init_context(struct ffmpeg_mux *ffm)
 {
 	AVOutputFormat *output_format;
 	int ret;
 	bool isNetwork = false;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: isNetwork -> is_network

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 7]
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
...
+	    strncmp(ffm->params.file, TCP_PROTO, sizeof(TCP_PROTO) - 1) == 0 ||
+	    isHTTP) {
 		isNetwork = true;
-
-	if (isNetwork) {
 		avformat_network_init();
+	}
+
+	if (isNetwork && !(isHTTP))
 		output_format = av_guess_format("mpegts", NULL, "video/M2PT");
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:542]: !is_http

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 8]
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
...
 
 static int ffmpeg_mux_init_context(struct ffmpeg_mux *ffm)
 {
 	AVOutputFormat *output_format;
 	int ret;
-	bool isNetwork = false;
+	bool is_network = false;
+	bool is_http = false;
+	is_http = (strncmp(ffm->params.file, HTTP_PROTO,
+			  sizeof(HTTP_PROTO) - 1) == 0);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Add a space before sizeof to align with ffm.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 9]
```diff
@@ -354,6 +360,50 @@ static bool ffmpeg_mux_start(void *data)
...
+	}
+
+	/* write headers and start capture */
+	os_atomic_set_bool(&stream->active, true);
+	os_atomic_set_bool(&stream->capturing, true);
+	stream->total_bytes = 0;
+	obs_output_begin_data_capture(stream->output, 0);
+
+	info("Writing to path '%s'...", stream->path.array);
+	dstr_free(&path);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Move dstr_free(&path); to after start_pipe(stream, path.array);

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 10]
```diff
@@ -354,6 +360,50 @@ static bool ffmpeg_mux_start(void *data)
...
+	const char *path_str;
+	const char *stream_key;
+	struct dstr path = {0};
+
+	if (!obs_output_can_begin_data_capture(stream->output, 0))
+		return false;
+	if (!obs_output_initialize_encoders(stream->output, 0))
+		return false;
+
+	settings = obs_output_get_settings(stream->output);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: We don't really need settings.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 11]
```diff
@@ -354,6 +360,50 @@ static bool ffmpeg_mux_start(void *data)
...
+	obs_service_t *service;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Move obs_service_t *service; to the beginning where all the variables are declared.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 12]
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
...
+	else
 		output_format = av_guess_format(NULL, ffm->params.file, NULL);
-	}
 
 	if (output_format == NULL) {
 		fprintf(stderr, "Couldn't find an appropriate muxer for '%s'\n",
 			ffm->params.file);
 		return FFM_ERROR;
 	}
+	printf("info: output_format name and long_name: %s, %s\n",
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:551]: Check if output_format->name or output_format->long_name is NULL. You can do something like:
output_format->name ? output_format->name : "unknown"

**Review by ushadow — CHANGES_REQUESTED** [review]
One more small change.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 13]
```diff
@@ -519,29 +519,39 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
...
+	else
 		output_format = av_guess_format(NULL, ffm->params.file, NULL);
-	}
 
 	if (output_format == NULL) {
 		fprintf(stderr, "Couldn't find an appropriate muxer for '%s'\n",
 			ffm->params.file);
 		return FFM_ERROR;
 	}
+	output_format->name ? output_format->name : "unknown";
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: These need to be in the printf statement.

printf("info: output_format name and long_name: %s, %s\n",
           output_format->name ? output_format->name : "unknown",
           output_format->long_name ? output_format->long_name : "unknown");

**Review by ushadow — CHANGES_REQUESTED** [review]
Please see the comment.

**Pull request merged**